kahuna/public/js/services/photoshoot.js

```diff
@@ -16,38 +16,38 @@ photoshootService.factory('photoshootService', [
         }
 
         function batchAdd({ data, images }) {
-            return trackAll($rootScope, "photoshoot", images, image =>
-                putPhotoshoot({data, image})
-            );
-        }
-
-        function batchRemove({ images }) {
-            return trackAll($rootScope, "photoshoot", images, image =>
-                deletePhotoshoot({ image })
-            );
-        }
-
-        function putPhotoshoot({ data, image }) {
-            return imageAccessor.getPhotoshoot(image)
-                .put({ data })
-                .then(newPhotoshoot => apiPoll(() => untilEqual({
+          return trackAll(
+            $rootScope,
+            "photoshoot",
+            images,
+            [
+              (image) => imageAccessor.getPhotoshoot(image).put({ data }),
+              (image, newPhotoshoot) =>
+                apiPoll(() =>
+                  untilEqual({
                     image,
                     expectedPhotoshoot: newPhotoshoot.data
-                })))
-                .then(newImage => {
-                    $rootScope.$emit('image-updated', newImage, image);
-                    return newImage;
-                });
+                  })
+                )
+            ],
+            "images-updated"
+          );
         }
 
-        function deletePhotoshoot({ image }) {
-            return imageAccessor.getPhotoshoot(image)
-                .delete()
-                .then(() => apiPoll(() => untilEqual({image, expectedPhotoshoot: undefined })))
-                .then(newImage => {
-                    $rootScope.$emit('image-updated', newImage, image);
-                    return newImage;
-                });
+        function batchRemove({ images }) {
+          return trackAll(
+            $rootScope,
+            "photoshoot",
+            images,
+            [
+              (image) => imageAccessor.getPhotoshoot(image).delete(),
```

Comment:
Wherever we're using `trackAll` in this way, I think it makes a lot of sense to be explicit as to what we're waiting for. For example (hopefully I have the semantics right!),

```js
function batchAdd({ data, images }) {
          const putPhotoshoot = (image) => { /* ... etc */ }
          const waitForPhotoshootInApi = (image, newPhotoshoot) => (image, newPhotoshoot) =>  { /* ... etc */ }
          return trackAll(
            $rootScope,
            "photoshoot",
            images,
            [putPhotoshoot, waitForPhotoshootInApi],
            "images-updated"
          );
        }
```

Reply:
yeah, this is sensible.